File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode232.ts
class Stack<T> {
    private items: T[] = [];

    isEmpty(): boolean {
        return this.items.length === 0;
    }

    add(elm: T | undefined): void {
        this.items.push(elm as T);
    }

    peek(): T | undefined {
        if (this.isEmpty()) {
            throw new Error("Stack is Empty.");
        } else {
            return this.items[this.items.length - 1];
        }
    }

    pop(): T | undefined {
        if (this.isEmpty()) {
            throw new Error("Stack is empty.");
        } else {
            return this.items.pop();
        }
    }
}

class MyQueue {
    stack1: Stack<number>;
    stack2: Stack<number>;

    constructor() {
        this.stack1 = new Stack<number>();
        this.stack2 = new Stack<number>();
    }

    push(x: number): void {
        if (this.stack2.isEmpty()) {
            this.stack2.add(x);
        } else {
            while (!this.stack2.isEmpty()) {
                this.stack1.add(this.stack2.pop());
            }
            this.stack2.add(x);
            while (!this.stack1.isEmpty()) {
                this.stack2.add(this.stack1.pop());
            }
        }
    }

    pop(): number {
        return this.stack2.pop() as number;
    }

    peek(): number {
        return this.stack2.peek() as number;
    }

    empty(): boolean {
        return this.stack2.isEmpty();
    }
}

File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode232.java
import java.util.Stack;

class MyQueue {
    Stack<Integer> stack1;
    Stack<Integer> stack2;

    MyQueue() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }

    public void push(int x) {
        if (stack2.isEmpty()) {
            stack2.add(x);
        } else {
            while (!stack2.isEmpty()) {
                stack1.add(stack2.pop());
            }
            stack2.add(x);
            while (!stack1.isEmpty()) {
                stack2.add(stack1.pop());
            }
        }
    }

    public int pop() {
        return this.stack2.pop();
    }

    public int peek() {
        return this.stack2.peek();
    }

    public boolean empty() {
        return this.stack2.isEmpty();
    }
}

public class Leetcode232 {
    public static void main(String[] args) {

    }
}
